### PR TITLE
Mat 1654 Improve checkpoints gas usage

### PR DESCRIPTION
**Pull request by imyourm8**
Refactored checkpointing reducing gas usage up to 50% for checkpointing. 
Value is extrapolated, depends on amount of validators, new approach scales much better with growing number of validators.

```
Some benchmarks. all runs after very first checkpoint, where some variables are being initialized.
Before:
229514 - 7 validators
295020 - 10 validators

After:
161555 - 7 validators, ~29.6% improvement
196156 - 10 validators, ~33.5% improvement

Simple math shows that improvement over 1 validator overhead is 50%. Which will be closer with more validators.
```
**Important remark that gas usage above is best case when validators do not restake or withdraw their rewards. In case they do, new approach gas usage will not increase while old one will grow.**

I've done 3 steps in order to improve gas usage drastically:
1. Moved updateRewards and comissionRate from VoucherShares into StakeManager. The way rewards are calculated is lazy, any sell/buy/restake/withdraw triggers calculation of accumulated rewards for delegators and validator.
2. Inlining ecrecovery and other careful optimizations
3. Removing rewards calculation in the checkpoint at all. Instead trigger calculations on any state change (all delegators operations + slash/withdraw/restake/stake/unstake) using approach used for calculating delegators' reward.

### CODE REVIEW

**Comment by imyourm8** [comment]
Don't mind failling tests, all of it due to revert message mismatch due to bytecode optimizations or because state must be tracked differently now since some state is lazy.

File: contracts/common/governance/Governable.sol [hunk 1]
```diff
@@ -10,10 +10,14 @@ contract Governable {
     }
 
     modifier onlyGovernance() {
+        _assertGovernance();
```
Comment (0xAshish) [inline contracts/common/governance/Governable.sol:13]: what's the difference?

Comment (imyourm8) [reply contracts/common/governance/Governable.sol:13]: Modifier copies it's content to the function. Instead of copying revert messages everywhere (which occupies precious 24kb) let's copy function call. Tiny more gas and huge bytecode savings

File: contracts/staking/stakeManager/StakeManager.sol [hunk 2]
```diff
@@ -479,79 +496,198 @@ contract StakeManager is IStakeManager, StakeManagerStorage {
         reward = Math.min(CHECKPOINT_REWARD, reward);
 
         uint256 _proposerBonus = reward.mul(proposerBonus).div(100);
-        Validator storage _proposer = validators[signerToValidator[proposer]];
+        uint256 proposerId = signerToValidator[proposer];
```
Comment (0xAshish) [inline contracts/staking/stakeManager/StakeManager.sol:498]: we need to move the proposer bonus to `_increaseRewardAndAssertConsensus` because proposer bonus must be deducted from the amount of the final rewards.

File: contracts/common/governance/Governable.sol (hunk 1 above)
Comment (0xAshish) [reply contracts/common/governance/Governable.sol:13]: yeah cool.

File: contracts/staking/stakeManager/StakeManager.sol [hunk 3]
```diff
@@ -479,79 +495,199 @@ contract StakeManager is IStakeManager, StakeManagerStorage {
         reward = Math.min(CHECKPOINT_REWARD, reward);
 
         uint256 _proposerBonus = reward.mul(proposerBonus).div(100);
-        Validator storage _proposer = validators[signerToValidator[proposer]];
-        if (_proposer.contractAddress != address(0x0)) {
-            IValidatorShare(_proposer.contractAddress).addProposerBonus(_proposerBonus, _proposer.amount);
+        uint256 proposerId = signerToValidator[proposer];
+        Validator storage _proposer = validators[proposerId];
```
Comment (0xAshish) [inline contracts/staking/stakeManager/StakeManager.sol:499]: we need to add proposer rewards into total rewards var.

File: contracts/staking/stakeManager/StakeManager.sol [hunk 4]
```diff
@@ -479,79 +495,199 @@ contract StakeManager is IStakeManager, StakeManagerStorage {
...
             _proposer.reward = _proposer.reward.add(_proposerBonus);
         }
 
-        reward = reward.sub(_proposerBonus);
-        uint256 stakePower = currentValidatorSetTotalStake();
         // update stateMerkleTree root for accounts balance on heimdall chain
         accountStateRoot = stateRoot;
-        _finalizeCommit();
-        return checkSignature(stakePower, reward, voteHash, sigs);
+        checkSignature(currentValidatorSetTotalStake(), reward.sub(_proposerBonus), voteHash, sigs);
```
Comment (0xAshish) [inline contracts/staking/stakeManager/StakeManager.sol:514]: @imyourm8 Move 490 to 513 in `_increaseRewardAndAssertConsensus`, previously it was design bottleneck due to rewards distribution now we can do it after.
- it will save gas in terms of failed tx.
- accurate rewards calculation with accurate 2/3 power.

File: contracts/staking/stakeManager/StakeManager.sol [hunk 5]
```diff
@@ -382,10 +384,23 @@ contract StakeManager is IStakeManager, StakeManagerStorage {
         factory = ValidatorShareFactory(newFactory);
     }
 
-    function updateValidatorState(uint256 validatorId, int256 amount) public {
-        require(validators[validatorId].contractAddress == msg.sender, "Invalid contract address");
+    function updateValidatorState(uint256 validatorId, int256 amount) public onlyDelegation(validatorId) {
         uint256 _currentEpoch = currentEpoch;
         validatorState[_currentEpoch].amount = (validatorState[_currentEpoch].amount + amount);
+        
+        if (amount >= 0) {
```
Comment (0xAshish) [inline contracts/staking/stakeManager/StakeManager.sol:391]: why this is required?

Comment (0xAshish) [reply contracts/staking/stakeManager/StakeManager.sol:391]: in case of jail/unjail?

Comment (coinsandsteeldev) [reply contracts/staking/stakeManager/StakeManager.sol:391]: Yes, will do >, < checks to avoid unnecessary call

File: contracts/staking/stakeManager/StakeManager.sol (hunk 4 above)
Comment (coinsandsteeldev) [reply contracts/staking/stakeManager/StakeManager.sol:514]: I don't get it, 490 is a comment line

File: contracts/staking/stakeManager/StakeManager.sol (hunk 3 above)
Comment (coinsandsteeldev) [reply contracts/staking/stakeManager/StakeManager.sol:499]: Where is total rewards? The one that was in the storage?

File: contracts/staking/stakeManager/StakeManager.sol (hunk 4 above)
Comment (coinsandsteeldev) [reply contracts/staking/stakeManager/StakeManager.sol:514]: You mean assert first before `rewardPerStake ` modification?

**Comment by 0xAshish** [comment]
overall looks good to me, need more eyes.

File: contracts/staking/stakeManager/StakeManager.sol (hunk 4 above)
Comment (0xAshish) [reply contracts/staking/stakeManager/StakeManager.sol:514]: I am talking about moving whole block 490-513.

File: contracts/staking/stakeManager/StakeManager.sol (hunk 5 above)
Comment (0xAshish) [reply contracts/staking/stakeManager/StakeManager.sol:391]: why are you updating the delegation amount in this function?

File: contracts/staking/stakeManager/StakeManager.sol (hunk 4 above)
Comment (coinsandsteeldev) [reply contracts/staking/stakeManager/StakeManager.sol:514]: got it

File: contracts/staking/stakeManager/StakeManager.sol (hunk 5 above)
Comment (coinsandsteeldev) [reply contracts/staking/stakeManager/StakeManager.sol:391]: Because before it was used everywhere as additional call next to `activeAmount` modifications. Now there is not activeAmount but `delegatedAmount` in the Validator struct. So i just moved that logic inside `updateValidatorState`

File: contracts/staking/stakeManager/StakeManager.sol (hunk 3 above)
Comment (0xAshish) [reply contracts/staking/stakeManager/StakeManager.sol:499]: yes

File: contracts/staking/stakeManager/StakeManager.sol [hunk 6]
```diff
@@ -479,79 +495,199 @@ contract StakeManager is IStakeManager, StakeManagerStorage {
...
             }
         }
 
-        reward = CHECKPOINT_REWARD.mul(totalStakePower).div(currentValidatorSetTotalStake());
-        totalRewards = totalRewards.add(reward);
-        require(totalStakePower >= currentValidatorSetTotalStake().mul(2).div(3).add(1), "2/3+1 non-majority!");
+        _increaseRewardAndAssertConsensus(checkpointStakePower, totalStakePower, reward);
+    }
+
+    function _increaseRewardAndAssertConsensus(uint256 checkpointStakePower, uint256 totalStakePower, uint256 reward) private {
```
Comment (jdkanani) [inline contracts/staking/stakeManager/StakeManager.sol:548]: Here `checkpointStakePower` is not being used.

**Comment by imyourm8** [comment]
After byte code compression work was done, StakeManager size was down to **22.96 kb** after optimizations.

**Comment by imyourm8** [comment]
Some results of benchmarks with 200 validators:

All validators signed - 2.35 million gas
Only 10 validators signed - 4.94 million gas

As the result even if almost nobody signs there is no risks of running out of gas and we have spare gas to increase validators set if necessary.

File: test/helpers/utils.js [hunk 7]
```diff
@@ -73,14 +73,20 @@ export function getSigsWithVotes(_wallets, data, sigPrefix, maxYesVotes) {
...
+  })
 }
 
-export async function checkPoint(wallets, proposer, stakeManager, { blockInterval = 1 } = {}) {
+export async function checkPoint(wallets, proposer, stakeManager, { blockInterval = 1, rootchainOwner } = {}) {
   const voteData = 'dummyData'
-  const sigs = ethUtils.bufferToHex(
-    encodeSigs(getSigs(wallets, ethUtils.keccak256(voteData)))
-  )
+  const sigs = encodeSigs(getSigs(wallets, ethUtils.keccak256(voteData)))
```
Comment (AyushKaul) [inline test/helpers/utils.js:88]: This is required for heimdall as well even pos portal

File: contracts/staking/stakeManager/ValidatorAuction.sol [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        uint256 currentAuctionAmount = auction.amount;
+
+        perceivedStake = Math.max(perceivedStake, currentAuctionAmount);
+
+        require(perceivedStake < amount, "Must bid higher");
+        require(token.transferFrom(msg.sender, address(this), amount), "Transfer failed");
+
+        //replace prev auction
+        if (currentAuctionAmount != 0) {
+            require(token.transfer(auction.user, currentAuctionAmount), "Bid return failed");
```
[the remark this answers is not in the document]
Comment (imyourm8) [reply contracts/staking/stakeManager/ValidatorAuction.sol:68]: Nothing i guess, we don't really take into account such edge cases.

File: test/helpers/utils.js (hunk 7 above)
Comment (imyourm8) [reply test/helpers/utils.js:88]: What is required?

Comment (AyushKaul) [reply test/helpers/utils.js:88]: encoding the sigs right?

Comment (imyourm8) [reply test/helpers/utils.js:88]: Oh yeah, i see, will fix this moment

File: contracts/staking/stakeManager/ValidatorAuction.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (AyushKaul) [reply contracts/staking/stakeManager/ValidatorAuction.sol:68]: But if recursively repeated than it can drain the funds in the contract. Ideally, if the bid is smaller than the gas then the bidder should not be refunded. This is not related to the pr i just thought, so pointing it out.

File: contracts/staking/StakingInfo.sol [hunk 9]
```diff
@@ -397,28 +400,24 @@ contract StakingInfo is Ownable {
...
-            _contract,
-            status
+            ,
+            status, , , ,delegatedAmount,
         ) = stakeManager.validators(validatorId);
-        _status = uint256(status);
-        if (_contract != address(0x0)) {
-            reward += IStakeManagerLocal(_contract).validatorRewards();
-        }
+        reward = reward.add(delegatedAmount);
```
Comment (jdkanani) [inline contracts/staking/StakingInfo.sol:420]: Is `delegatedAmount` the same as `reward`? `delegatedAmount` is total delegation for that user, right?

File: contracts/staking/stakeManager/StakeManager.sol [hunk 10]
```diff
@@ -102,45 +124,200 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable {
...
+    // New implementation upgrade
+
+    function migrateValidatorsData(
+        uint256 validatorIdFrom,
+        uint256 validatorIdTo
+    ) public onlyOwner {
+        for (uint256 i = validatorIdFrom; i < validatorIdTo; ++i) {
+            ValidatorShare contractAddress = ValidatorShare(validators[i].contractAddress);
+            if (contractAddress != ValidatorShare(0)) {
+                validators[i].delegatorsReward = contractAddress.validatorRewards_deprecated();
```
Comment (jdkanani) [inline contracts/staking/stakeManager/StakeManager.sol:221]: No need add `INITIALIZED_AMOUNT` to `validatorRewards` or `delegatedAmount`?

File: contracts/staking/stakeManager/StakeManager.sol [hunk 11]
```diff
@@ -102,45 +124,200 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable {
...
             keccak256(abi.encode(msg.sender, accumFeeAmount)).checkMembership(index, accountStateRoot, proof),
             "Wrong acc proof"
         );
         uint256 withdrawAmount = accumFeeAmount.sub(userFeeExit[msg.sender]);
         _claimFee(msg.sender, withdrawAmount);
         userFeeExit[msg.sender] = accumFeeAmount;
-        require(token.transfer(msg.sender, withdrawAmount));
-    }
-
-    function stake(
```
Comment (jdkanani) [inline contracts/staking/stakeManager/StakeManager.sol:137]: No `stake` function required? We can use `stakeFor`?

File: contracts/staking/stakeManager/StakeManagerStorage.sol [hunk 12]
```diff
@@ -82,4 +93,7 @@ contract StakeManagerStorage is GovernanceLockable, RootChainable {
     mapping(uint256 => uint256) public latestSignerUpdateEpoch;
 
     uint256 public totalHeimdallFee;
+
+    uint256 public rewardPerStake;
+    address auctionImplementation;
```
Comment (jdkanani) [inline contracts/staking/stakeManager/StakeManagerStorage.sol:98]: Shouldn't this mess with `StakeManagerStorage, Initializable` order for `StakeManager` since `Initializable` also has storage?

File: contracts/staking/validatorShare/ValidatorShare.sol [hunk 13]
```diff
@@ -146,89 +85,48 @@ contract ValidatorShare is IValidatorShare, ERC20NonTransferable, OwnableLockabl
...
-        uint256 rate = exchangeRate();
-        if (shares == 0) {
-            return (0, rate);
-        }
+    /**
+        Public Methods
+     */
 
-        return (rate.mul(shares).div(_getRatePrecision()), rate);
+    function buyVoucher(uint256 _amount, uint256 _minSharesToMint) public {
```
Comment (jdkanani) [inline contracts/staking/validatorShare/ValidatorShare.sol:103]: It's good to have `amountToDeposit` as return value to `buyVoucher`. Other contracts can use that value.

**Review by jdkanani — CHANGES_REQUESTED** [review]

File: contracts/staking/StakingInfo.sol [hunk 14]
```diff
@@ -20,6 +18,11 @@ contract IStakeManagerLocal {
         address signer;
         address contractAddress;
         Status status;
+        uint256 commissionRate;
+        uint256 lastCommissionUpdate;
+        uint256 accumulatedReward;
```
Comment (jdkanani) [inline contracts/staking/StakingInfo.sol:23]: Is it the same as `delegatorsReward`? I see two different names.

**Review by jdkanani — CHANGES_REQUESTED** [review]

File: contracts/staking/stakeManager/StakeManager.sol [hunk 15]
```diff
@@ -491,123 +532,135 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable {
...
-                        checkpointStakePower,
-                        validatorStake(validatorId)
-                    );
-                } else {
-                    valPow = validator.amount;
-                    validator.reward = validator.reward.add(valPow.mul(reward).div(checkpointStakePower));
+                signedStakePower = signedStakePower.add(validators[validatorId].delegatedAmount.add(amount));
+
+                if (unstakeCtx.deactivationEpoch != 0) {
+                    unstakeCtx.deactivatedValidators[unstakeCtx.validatorIndex] = validatorId;
```
Comment (jdkanani) [inline contracts/staking/stakeManager/StakeManager.sol:590]: Do we need to handle this explicitly? Because if it is a "leaving" validator and still active, shouldn't it be the same as an active validator?

File: contracts/staking/stakeManager/StakeManager.sol [hunk 16]
```diff
@@ -631,38 +684,288 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable {
...
+        );
+
+        // evaluate rewards for validator who did't sign and set latest reward per stake to new value to avoid them from getting new rewards.
+        _updateValidatorsRewards(unsignedValidators, totalUnsignedValidators, newRewardPerStake);
+
+        // distribute rewards between signed validators
+        rewardPerStake = newRewardPerStake;
+
+        // evaluate rewards for unstaked validators to avoid getting new rewards until they claim their stake
+        _updateValidatorsRewards(deactivatedValidators, totalDeactivatedValidators, newRewardPerStake);
```
Comment (jdkanani) [inline contracts/staking/stakeManager/StakeManager.sol:833]: Shouldn't `_updateValidatorsRewards` before `rewardsPerStake` or after? 

`_updateValidatorsRewards` for both `unsigned` and `deactivated`.

**Review by jdkanani — CHANGES_REQUESTED** [review]

File: contracts/staking/StakingInfo.sol (hunk 9 above)
Comment (imyourm8) [reply contracts/staking/StakingInfo.sol:420]: I am not sure why is it here. Fixed.

File: contracts/staking/stakeManager/StakeManager.sol (hunk 10 above)
Comment (imyourm8) [reply contracts/staking/stakeManager/StakeManager.sol:221]: Fixed.

File: contracts/staking/stakeManager/StakeManager.sol (hunk 11 above)
Comment (imyourm8) [reply contracts/staking/stakeManager/StakeManager.sol:137]: Correct.

File: contracts/staking/validatorShare/ValidatorShare.sol (hunk 13 above)
Comment (imyourm8) [reply contracts/staking/validatorShare/ValidatorShare.sol:103]: Done.

File: contracts/staking/StakingInfo.sol (hunk 14 above)
Comment (imyourm8) [reply contracts/staking/StakingInfo.sol:23]: Fixed.

File: contracts/staking/stakeManager/StakeManager.sol (hunk 15 above)
Comment (imyourm8) [reply contracts/staking/stakeManager/StakeManager.sol:590]: Not really, we must explicitly update his rewards, otherwise, he will not get rewards. UnstakeClaim doesn't update rewards to avoid unstaked validator from getting more rewards.

File: contracts/staking/stakeManager/StakeManager.sol (hunk 16 above)
Comment (imyourm8) [reply contracts/staking/stakeManager/StakeManager.sol:833]: For unstaked - no, they must get the last reward.

File: contracts/staking/stakeManager/StakeManagerStorage.sol (hunk 12 above)
Comment (imyourm8) [reply contracts/staking/stakeManager/StakeManagerStorage.sol:98]: It is going to mess, yes, I am looking into it.

File: contracts/staking/stakeManager/StakeManager.sol [hunk 17]
```diff
@@ -631,38 +683,288 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable {
...
+                validatorsStake,
+                delegatedAmount,
+                _getEligibleValidatorReward(validatorId, combinedStakePower, currentRewardPerStake)
+            );
+        } else {
+            _increaseValidatorReward(
+                validatorId,
+                _getEligibleValidatorReward(validatorId, validatorsStake, currentRewardPerStake)
+            );
+        }
```
Comment (jdkanani) [inline contracts/staking/stakeManager/StakeManager.sol:869]: Can't we just skip this code block if `currentRewardPerStake <= rewardsPerStake`? And skip directly to `validators[validatorId].initialRewardPerStake = newRewardPerStake`?

**Review by jdkanani — CHANGES_REQUESTED** [review]

File: contracts/staking/stakeManager/StakeManagerStorageExtension.sol [hunk 18]
```diff
@@ -0,0 +1,7 @@
+pragma solidity 0.5.17;
+
+contract StakeManagerStorageExtension {
+    uint256 public rewardPerStake;
```
[the remark this answers is not in the document]
Comment (imyourm8) [reply contracts/staking/stakeManager/StakeManagerStorageExtension.sol:4]: No, it's going to work just fine.

File: contracts/staking/stakeManager/StakeManager.sol [hunk 19]
```diff
@@ -102,45 +124,201 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable {
...
+    }
+
+    function delegatorsReward(uint256 validatorId) public view returns (uint256) {
+        (, uint256 _delegatorsReward) = _evaluateValidatorAndDelegationReward(validatorId);
+        return validators[validatorId].delegatorsReward.add(_delegatorsReward).sub(INITIALIZED_AMOUNT);
+    }
+
+    function validatorReward(uint256 validatorId) public view returns (uint256) {
+        uint256 _validatorReward;
+        if (validators[validatorId].deactivationEpoch == 0) {
```
[the remark this answers is not in the document]
Comment (imyourm8) [reply contracts/staking/stakeManager/StakeManager.sol:142]: Totally, otherwise, UI will show ever-increasing reward.

[the remark this answers is not in the document]
Comment (jdkanani) [reply contracts/staking/stakeManager/StakeManager.sol:142]: Then shouldn't it be `validators[validatorId].deactivationEpoch < currentEpoch`?

File: contracts/staking/stakeManager/StakeManager.sol (hunk 17 above)
Comment (imyourm8) [reply contracts/staking/stakeManager/StakeManager.sol:869]: I can skip it yes, will do.

File: contracts/staking/stakeManager/StakeManager.sol [hunk 20]
```diff
@@ -231,6 +232,10 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable, Sign
         signers = _signers;
     }
 
+    function fixStorage() public onlyOwner {
+
+    }
```
[the remark this answers is not in the document]
Comment (imyourm8) [reply contracts/staking/stakeManager/StakeManager.sol]: Already removed

**Review by jdkanani — APPROVED** [review]
